<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
<meta name="generator" content="pdoc3 0.11.5">
<title>midsv.format API documentation</title>
<meta name="description" content="">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/sanitize.min.css" integrity="sha512-y1dtMcuvtTMJc1yPgEqF0ZjQbhnc/bFhyvIyVNb9Zk5mIGtqVaAB1Ttl28su8AvFMOY0EwRbAe+HCLqj6W7/KA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/13.0.0/typography.min.css" integrity="sha512-Y1DYSb995BAfxobCkKepB1BqJJTPrOp3zPL74AWFugHHmmdcvO+C48WLrUOlhGMc0QG7AE3f7gmvvcrmX2fDoA==" crossorigin>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/styles/default.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:1.5em;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:2em 0 .50em 0}h3{font-size:1.4em;margin:1.6em 0 .7em 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .2s ease-in-out}a:visited{color:#503}a:hover{color:#b62}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900;font-weight:bold}pre code{font-size:.8em;line-height:1.4em;padding:1em;display:block}code{background:#f3f3f3;font-family:"DejaVu Sans Mono",monospace;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source > summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible;min-width:max-content}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em 1em;margin:1em 0}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul ul{padding-left:1em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.9.0/highlight.min.js" integrity="sha512-D9gUyxqja7hBtkWpPWGt9wfbfaMGVt9gnyCvYa+jojwwPHLCzUm5i8rpk7vD7wNee9bA35eYIjobYPaQuKS1MQ==" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => {
hljs.configure({languages: ['bash', 'css', 'diff', 'graphql', 'ini', 'javascript', 'json', 'plaintext', 'python', 'python-repl', 'rust', 'shell', 'sql', 'typescript', 'xml', 'yaml']});
hljs.highlightAll();
/* Collapse source docstrings */
setTimeout(() => {
[...document.querySelectorAll('.hljs.language-python > .hljs-string')]
.filter(el => el.innerHTML.length > 200 && ['"""', "'''"].includes(el.innerHTML.substring(0, 3)))
.forEach(el => {
let d = document.createElement('details');
d.classList.add('hljs-string');
d.innerHTML = '<summary>"""</summary>' + el.innerHTML.substring(3);
el.replaceWith(d);
});
}, 100);
})</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>midsv.format</code></h1>
</header>
<section id="section-intro">
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="midsv.format.dictionarize_sam"><code class="name flex">
<span>def <span class="ident">dictionarize_sam</span></span>(<span>sam: list[list[str]]) ‑> list[dict[str | int]]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dictionarize_sam(sam: list[list[str]]) -&gt; list[dict[str | int]]:
    &#34;&#34;&#34;Extract mapped alignments from SAM

    Args:
        sam (list[list]): a list of lists of SAM format including CS tag

    Returns:
        dict: a dictionary containing QNAME, RNAME, POS, QUAL, CSTAG and RLEN
    &#34;&#34;&#34;
    aligns = []
    for alignment in sam:
        if alignment[0].startswith(&#34;@&#34;):
            continue
        if alignment[2] == &#34;*&#34;:
            continue
        if alignment[9] == &#34;*&#34;:
            continue
        for i, a in enumerate(alignment):
            if a.startswith(&#34;cs:Z:&#34;) and not re.search(r&#34;:[0-9]+&#34;, alignment[i]):
                idx_cstag = i
        alignments = dict(
            QNAME=alignment[0].replace(&#34;,&#34;, &#34;_&#34;),
            FLAG=int(alignment[1]),
            RNAME=alignment[2],
            POS=int(alignment[3]),
            CIGAR=alignment[5],
            SEQ=alignment[9],
            QUAL=alignment[10],
            CSTAG=alignment[idx_cstag],
        )
        aligns.append(alignments)
    aligns = sorted(aligns, key=lambda x: [x[&#34;QNAME&#34;], x[&#34;POS&#34;]])
    return aligns</code></pre>
</details>
<div class="desc"><p>Extract mapped alignments from SAM</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>a list of lists of SAM format including CS tag</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>dict</code></dt>
<dd>a dictionary containing QNAME, RNAME, POS, QUAL, CSTAG and RLEN</dd>
</dl></div>
</dd>
<dt id="midsv.format.extract_sqheaders"><code class="name flex">
<span>def <span class="ident">extract_sqheaders</span></span>(<span>sam: list[list]) ‑> dict[str, int]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_sqheaders(sam: list[list]) -&gt; dict[str, int]:
    &#34;&#34;&#34;Extract SN (Reference sequence name) and LN (Reference sequence length) from SQ header

    Args:
        sam (list[list]): a list of lists of SAM format

    Returns:
        dict: a dictionary containing (multiple) SN and LN
    &#34;&#34;&#34;
    sqheaders = [s for s in sam if &#34;@SQ&#34; in s]
    SNLN = {}
    for sqheader in sqheaders:
        sn_ln = [sq for sq in sqheader if re.search((&#34;SN:|LN:&#34;), sq)]
        sn = sn_ln[0].replace(&#34;SN:&#34;, &#34;&#34;)
        ln = sn_ln[1].replace(&#34;LN:&#34;, &#34;&#34;)
        SNLN.update({sn: int(ln)})
    return SNLN</code></pre>
</details>
<div class="desc"><p>Extract SN (Reference sequence name) and LN (Reference sequence length) from SQ header</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>a list of lists of SAM format</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>dict</code></dt>
<dd>a dictionary containing (multiple) SN and LN</dd>
</dl></div>
</dd>
<dt id="midsv.format.realign_sequence"><code class="name flex">
<span>def <span class="ident">realign_sequence</span></span>(<span>alignment: dict[str, str | int]) ‑> dict</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def realign_sequence(alignment: dict[str, str | int]) -&gt; dict:
    &#34;&#34;&#34;Discard insertion, and add deletion and spliced nucleotides to unify sequence length&#34;&#34;&#34;
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    sequence = alignment[&#34;SEQ&#34;]
    sequence_ignored = [&#34;N&#34;] * alignment[&#34;POS&#34;]
    start = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig:
            end = start + int(cig[:-1])
            sequence_ignored.append(sequence[start:end])
            start = end
        elif &#34;I&#34; in cig:
            start += int(cig[:-1])
        elif any(x in cig for x in [&#34;D&#34;, &#34;N&#34;]):
            sequence_ignored.append(&#34;N&#34; * int(cig[:-1]))
    realignment = deepcopy(alignment)
    realignment[&#34;SEQ&#34;] = &#34;&#34;.join(sequence_ignored)
    return realignment</code></pre>
</details>
<div class="desc"><p>Discard insertion, and add deletion and spliced nucleotides to unify sequence length</p></div>
</dd>
<dt id="midsv.format.remove_resequence"><code class="name flex">
<span>def <span class="ident">remove_resequence</span></span>(<span>alignments: list[dict[str, str | int]]) ‑> list[dict]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def remove_resequence(alignments: list[dict[str, str | int]]) -&gt; list[dict]:
    &#34;&#34;&#34;Remove non-microhomologic overlapped reads within the same QNAME.
    The overlapped sequences can be (1) realignments by microhomology or (2) resequence by sequencing error.
    The &#39;realignments&#39; is not sequencing errors, and it preserves the same sequence.
    In contrast, the &#39;resequence&#39; is a sequencing error with the following characteristics:
    (1) The shorter reads that are completely included in the longer reads
    (2) Overlapped but not the same DNA sequence
    The resequenced fragments will be discarded and the longest alignment will be retain.
    Example reads are in `tests/data/overlap/real_overlap.sam` and `tests/data/overlap/real_overlap2.sam`

    Args:
        alignments (list[dict[str, str | int]]): disctionalized alignments

    Returns:
        list[list]: disctionalized SAM with removed overlaped reads
    &#34;&#34;&#34;
    alignments.sort(key=lambda x: x[&#34;QNAME&#34;])
    sam_groupby = groupby(alignments, lambda x: x[&#34;QNAME&#34;])
    sam_nonoverlapped = []
    for _, alignments in sam_groupby:
        alignments = list(alignments)
        if len(alignments) == 1:
            sam_nonoverlapped += alignments
            continue
        alignments = [realign_sequence(alignment) for alignment in alignments]
        alignments = sorted(alignments, key=lambda x: [x[&#34;POS&#34;]])
        is_overraped = False
        end_of_previous_read = -1
        previous_read = alignments[0][&#34;SEQ&#34;]
        for i, alignment in enumerate(alignments):
            if i == 0:
                start_of_previous_read = alignment[&#34;POS&#34;] - 1
                end_of_previous_read = return_end_of_current_read(alignment)
                continue
            start_of_current_read = alignment[&#34;POS&#34;] - 1
            end_of_current_read = return_end_of_current_read(alignment)
            # (1) The shorter reads that are completely included in the longer reads
            if start_of_previous_read &lt;= start_of_current_read and end_of_previous_read &gt;= end_of_current_read:
                is_overraped = True
                break
            else:
                start_overlap = max(start_of_previous_read, start_of_current_read)
                end_overlap = min(end_of_previous_read, end_of_current_read)
                for prev, curr in zip(
                    previous_read[start_overlap:end_overlap], alignment[&#34;SEQ&#34;][start_overlap:end_overlap]
                ):
                    if prev == &#34;N&#34; or curr == &#34;N&#34;:
                        continue
                    # (2) Overlapped but not the same DNA sequence
                    if prev != curr:
                        is_overraped = True
                        break
            start_of_previous_read = start_of_current_read
            end_of_previous_read = return_end_of_current_read(alignment)
        if is_overraped:
            # The longest alignment will be retain
            alignment = sorted(alignments, key=lambda x: -len(x[&#34;SEQ&#34;]))[0]
            sam_nonoverlapped.append(alignment)
        else:
            sam_nonoverlapped += alignments
    return sam_nonoverlapped</code></pre>
</details>
<div class="desc"><p>Remove non-microhomologic overlapped reads within the same QNAME.
The overlapped sequences can be (1) realignments by microhomology or (2) resequence by sequencing error.
The 'realignments' is not sequencing errors, and it preserves the same sequence.
In contrast, the 'resequence' is a sequencing error with the following characteristics:
(1) The shorter reads that are completely included in the longer reads
(2) Overlapped but not the same DNA sequence
The resequenced fragments will be discarded and the longest alignment will be retain.
Example reads are in <code>tests/data/overlap/real_overlap.sam</code> and <code>tests/data/overlap/real_overlap2.sam</code></p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>alignments</code></strong> :&ensp;<code>list[dict[str, str | int]]</code></dt>
<dd>disctionalized alignments</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>list[list]</code></dt>
<dd>disctionalized SAM with removed overlaped reads</dd>
</dl></div>
</dd>
<dt id="midsv.format.remove_softclips"><code class="name flex">
<span>def <span class="ident">remove_softclips</span></span>(<span>alignments: list[dict[str | int]]) ‑> list[dict[str | int]]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def remove_softclips(alignments: list[dict[str | int]]) -&gt; list[dict[str | int]]:
    &#34;&#34;&#34;Remove softclip information from SEQ and QUAL.

    Args:
        alignments (list[dict[str, str | int]]): disctionalized alignments

    Returns:
        list[list]: disctionalized SAM with trimmed softclips in QUAL
    &#34;&#34;&#34;
    alignments_softclips_removed = []
    for alignment in alignments:
        cigar = alignment[&#34;CIGAR&#34;]
        if &#34;S&#34; not in cigar:
            alignments_softclips_removed.append(alignment)
            continue
        cigar_split = split_cigar(cigar)
        left, right = cigar_split[0], cigar_split[-1]
        if &#34;S&#34; in left:
            left = int(left[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][left:]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][left:]
        if &#34;S&#34; in right:
            right = int(right[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][:-right]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][:-right]
        alignments_softclips_removed.append(alignment)
    return alignments_softclips_removed</code></pre>
</details>
<div class="desc"><p>Remove softclip information from SEQ and QUAL.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>alignments</code></strong> :&ensp;<code>list[dict[str, str | int]]</code></dt>
<dd>disctionalized alignments</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>list[list]</code></dt>
<dd>disctionalized SAM with trimmed softclips in QUAL</dd>
</dl></div>
</dd>
<dt id="midsv.format.return_end_of_current_read"><code class="name flex">
<span>def <span class="ident">return_end_of_current_read</span></span>(<span>alignment: dict[str, str | int]) ‑> int</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def return_end_of_current_read(alignment: dict[str, str | int]) -&gt; int:
    start_of_current_read = alignment[&#34;POS&#34;]
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    alignment_length = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig or &#34;D&#34; in cig or &#34;N&#34; in cig:
            alignment_length += int(cig[:-1])
    return start_of_current_read + alignment_length - 1</code></pre>
</details>
<div class="desc"></div>
</dd>
<dt id="midsv.format.split_cigar"><code class="name flex">
<span>def <span class="ident">split_cigar</span></span>(<span>cigar: str) ‑> list[str]</span>
</code></dt>
<dd>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def split_cigar(cigar: str) -&gt; list[str]:
    cigar_iter = iter(re.split(r&#34;([MIDNSHPX=])&#34;, cigar))
    cigar_splitted = [i + op for i, op in zip(cigar_iter, cigar_iter)]
    return cigar_splitted</code></pre>
</details>
<div class="desc"></div>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="midsv" href="index.html">midsv</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="midsv.format.dictionarize_sam" href="#midsv.format.dictionarize_sam">dictionarize_sam</a></code></li>
<li><code><a title="midsv.format.extract_sqheaders" href="#midsv.format.extract_sqheaders">extract_sqheaders</a></code></li>
<li><code><a title="midsv.format.realign_sequence" href="#midsv.format.realign_sequence">realign_sequence</a></code></li>
<li><code><a title="midsv.format.remove_resequence" href="#midsv.format.remove_resequence">remove_resequence</a></code></li>
<li><code><a title="midsv.format.remove_softclips" href="#midsv.format.remove_softclips">remove_softclips</a></code></li>
<li><code><a title="midsv.format.return_end_of_current_read" href="#midsv.format.return_end_of_current_read">return_end_of_current_read</a></code></li>
<li><code><a title="midsv.format.split_cigar" href="#midsv.format.split_cigar">split_cigar</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.11.5</a>.</p>
</footer>
</body>
</html>
