<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>midsv.format API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>midsv.format</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from __future__ import annotations
import re
from itertools import groupby
from copy import deepcopy

###########################################################
# Format headers and alignments
###########################################################


def extract_sqheaders(sam: list[list]) -&gt; dict[str, int]:
    &#34;&#34;&#34;Extract SN (Reference sequence name) and LN (Reference sequence length) from SQ header

    Args:
        sam (list[list]): a list of lists of SAM format

    Returns:
        dict: a dictionary containing (multiple) SN and LN
    &#34;&#34;&#34;
    sqheaders = [s for s in sam if &#34;@SQ&#34; in s]
    SNLN = {}
    for sqheader in sqheaders:
        sn_ln = [sq for sq in sqheader if re.search((&#34;SN:|LN:&#34;), sq)]
        sn = sn_ln[0].replace(&#34;SN:&#34;, &#34;&#34;)
        ln = sn_ln[1].replace(&#34;LN:&#34;, &#34;&#34;)
        SNLN.update({sn: int(ln)})
    return SNLN


def dictionarize_sam(sam: list[list]) -&gt; list[dict]:
    &#34;&#34;&#34;Extract mapped alignments from SAM

    Args:
        sam (list[list]): a list of lists of SAM format including CS tag

    Returns:
        dict: a dictionary containing QNAME, RNAME, POS, QUAL, CSTAG and RLEN
    &#34;&#34;&#34;
    aligns = []
    for alignment in sam:
        if alignment[0].startswith(&#34;@&#34;):
            continue
        if alignment[2] == &#34;*&#34;:
            continue
        if alignment[9] == &#34;*&#34;:
            continue
        for i, a in enumerate(alignment):
            if a.startswith(&#34;cs:Z:&#34;) and not re.search(r&#34;:[0-9]+&#34;, alignment[i]):
                idx_cstag = i
        samdict = dict(
            QNAME=alignment[0].replace(&#34;,&#34;, &#34;_&#34;),
            FLAG=int(alignment[1]),
            RNAME=alignment[2],
            POS=int(alignment[3]),
            CIGAR=alignment[5],
            SEQ=alignment[9],
            QUAL=alignment[10],
            CSTAG=alignment[idx_cstag],
        )
        aligns.append(samdict)
    aligns = sorted(aligns, key=lambda x: [x[&#34;QNAME&#34;], x[&#34;POS&#34;]])
    return aligns


###########################################################
# Remove undesired reads
###########################################################


def split_cigar(cigar: str) -&gt; list[str]:
    cigar_iter = iter(re.split(r&#34;([MIDNSHPX=])&#34;, cigar))
    cigar_splitted = [i + op for i, op in zip(cigar_iter, cigar_iter)]
    return cigar_splitted


def remove_softclips(sam: list[dict]) -&gt; list[dict]:
    &#34;&#34;&#34;Remove softclip information from SEQ and QUAL.

    Args:
        sam (list[list]): disctionalized SAM

    Returns:
        list[list]: disctionalized SAM with trimmed softclips in QUAL
    &#34;&#34;&#34;
    sam_list = []
    for alignment in sam:
        cigar = alignment[&#34;CIGAR&#34;]
        if &#34;S&#34; not in cigar:
            sam_list.append(alignment)
            continue
        cigar_split = split_cigar(cigar)
        left, right = cigar_split[0], cigar_split[-1]
        if &#34;S&#34; in left:
            left = int(left[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][left:]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][left:]
        if &#34;S&#34; in right:
            right = int(right[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][:-right]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][:-right]
        sam_list.append(alignment)
    return sam_list


def return_end_of_current_read(alignment:dict) -&gt; int:
    start_of_current_read = alignment[&#34;POS&#34;]
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    alignment_length = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig or &#34;D&#34; in cig or &#34;N&#34; in cig:
            alignment_length += int(cig[:-1])
    return start_of_current_read + alignment_length - 1


def realign_sequence(alignment: dict) -&gt; dict:
    &#34;&#34;&#34;Discard insertion, and add deletion and spliced nucreotide to unify sequence length
    &#34;&#34;&#34;
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    sequence = alignment[&#34;SEQ&#34;]
    sequence_ignored = [&#34;N&#34;] * alignment[&#34;POS&#34;]
    start = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig:
            end = start + int(cig[:-1])
            sequence_ignored.append(sequence[start: end])
            start = end
        elif &#34;I&#34; in cig:
            start += int(cig[:-1])
        elif any(x in cig for x in [&#34;D&#34;, &#34;N&#34;]):
            sequence_ignored.append(&#34;N&#34; * int(cig[:-1]))
    realignment = deepcopy(alignment)
    realignment[&#34;SEQ&#34;] = &#34;&#34;.join(sequence_ignored)
    return realignment



def remove_resequence(samdict: list[list]) -&gt; list[list]:
    &#34;&#34;&#34;Remove non-microhomologic overlapped reads within the same QNAME.
    The overlapped sequences can be (1) realignments by microhomology or (2) resequence by sequencing error.
    The resequenced reads are overlapped but not the same DNA sequence. In contrast, realignments preserve the same sequence.
    Since resequence is a Nanopore sequencing error, the reads should be discarded.
    In addition, if the start/end of two reads are exactly same, these sequences are resequenced so discarted.
    Example reads are in `tests/data/overlap/real_overlap.sam` and `tests/data/overlap/real_overlap2.sam`

    Args:
        sam (list[list]): disctionalized SAM

    Returns:
        list[list]: disctionalized SAM with removed overlaped reads
    &#34;&#34;&#34;
    samdict.sort(key=lambda x: x[&#34;QNAME&#34;])
    sam_groupby = groupby(samdict, lambda x: x[&#34;QNAME&#34;])
    sam_nonoverlapped = []
    for _, alignments in sam_groupby:
        alignments = list(alignments)
        if len(alignments) == 1:
            sam_nonoverlapped += alignments
            continue
        alignments = [realign_sequence(alignment) for alignment in alignments]
        alignments = sorted(alignments, key=lambda x: [x[&#34;POS&#34;], len(x[&#34;SEQ&#34;])]) # real_overlap2 needs comparisons in the following order: (1) shortest read (2) longest read ,(3) middle read
        is_overraped = False
        end_of_previous_read = -1
        previous_read = alignments[0][&#34;SEQ&#34;]
        for alignment in alignments:
            start_of_current_read = alignment[&#34;POS&#34;] - 1
            if end_of_previous_read &gt; start_of_current_read:
                end_of_current_read = return_end_of_current_read(alignment)
                start_overlap = max(start_of_previous_read, start_of_current_read)
                end_overlap = min(end_of_previous_read, end_of_current_read)
                if start_of_previous_read == start_of_current_read:
                    is_overraped = True
                    break
                if end_of_previous_read == end_of_current_read:
                    is_overraped = True
                    break
                for prev, curr in zip(previous_read[start_overlap: end_overlap], alignment[&#34;SEQ&#34;][start_overlap: end_overlap]):
                    if prev == &#34;N&#34; or curr == &#34;N&#34;:
                        continue
                    if prev != curr:
                        is_overraped = True
                        break
            end_of_previous_read = return_end_of_current_read(alignment)
            start_of_previous_read = start_of_current_read
        if not is_overraped:
            sam_nonoverlapped += alignments
    return sam_nonoverlapped</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="midsv.format.dictionarize_sam"><code class="name flex">
<span>def <span class="ident">dictionarize_sam</span></span>(<span>sam: list[list]) ‑> list[dict]</span>
</code></dt>
<dd>
<div class="desc"><p>Extract mapped alignments from SAM</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>a list of lists of SAM format including CS tag</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>dict</code></dt>
<dd>a dictionary containing QNAME, RNAME, POS, QUAL, CSTAG and RLEN</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dictionarize_sam(sam: list[list]) -&gt; list[dict]:
    &#34;&#34;&#34;Extract mapped alignments from SAM

    Args:
        sam (list[list]): a list of lists of SAM format including CS tag

    Returns:
        dict: a dictionary containing QNAME, RNAME, POS, QUAL, CSTAG and RLEN
    &#34;&#34;&#34;
    aligns = []
    for alignment in sam:
        if alignment[0].startswith(&#34;@&#34;):
            continue
        if alignment[2] == &#34;*&#34;:
            continue
        if alignment[9] == &#34;*&#34;:
            continue
        for i, a in enumerate(alignment):
            if a.startswith(&#34;cs:Z:&#34;) and not re.search(r&#34;:[0-9]+&#34;, alignment[i]):
                idx_cstag = i
        samdict = dict(
            QNAME=alignment[0].replace(&#34;,&#34;, &#34;_&#34;),
            FLAG=int(alignment[1]),
            RNAME=alignment[2],
            POS=int(alignment[3]),
            CIGAR=alignment[5],
            SEQ=alignment[9],
            QUAL=alignment[10],
            CSTAG=alignment[idx_cstag],
        )
        aligns.append(samdict)
    aligns = sorted(aligns, key=lambda x: [x[&#34;QNAME&#34;], x[&#34;POS&#34;]])
    return aligns</code></pre>
</details>
</dd>
<dt id="midsv.format.extract_sqheaders"><code class="name flex">
<span>def <span class="ident">extract_sqheaders</span></span>(<span>sam: list[list]) ‑> dict[str, int]</span>
</code></dt>
<dd>
<div class="desc"><p>Extract SN (Reference sequence name) and LN (Reference sequence length) from SQ header</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>a list of lists of SAM format</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>dict</code></dt>
<dd>a dictionary containing (multiple) SN and LN</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_sqheaders(sam: list[list]) -&gt; dict[str, int]:
    &#34;&#34;&#34;Extract SN (Reference sequence name) and LN (Reference sequence length) from SQ header

    Args:
        sam (list[list]): a list of lists of SAM format

    Returns:
        dict: a dictionary containing (multiple) SN and LN
    &#34;&#34;&#34;
    sqheaders = [s for s in sam if &#34;@SQ&#34; in s]
    SNLN = {}
    for sqheader in sqheaders:
        sn_ln = [sq for sq in sqheader if re.search((&#34;SN:|LN:&#34;), sq)]
        sn = sn_ln[0].replace(&#34;SN:&#34;, &#34;&#34;)
        ln = sn_ln[1].replace(&#34;LN:&#34;, &#34;&#34;)
        SNLN.update({sn: int(ln)})
    return SNLN</code></pre>
</details>
</dd>
<dt id="midsv.format.realign_sequence"><code class="name flex">
<span>def <span class="ident">realign_sequence</span></span>(<span>alignment: dict) ‑> dict</span>
</code></dt>
<dd>
<div class="desc"><p>Discard insertion, and add deletion and spliced nucreotide to unify sequence length</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def realign_sequence(alignment: dict) -&gt; dict:
    &#34;&#34;&#34;Discard insertion, and add deletion and spliced nucreotide to unify sequence length
    &#34;&#34;&#34;
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    sequence = alignment[&#34;SEQ&#34;]
    sequence_ignored = [&#34;N&#34;] * alignment[&#34;POS&#34;]
    start = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig:
            end = start + int(cig[:-1])
            sequence_ignored.append(sequence[start: end])
            start = end
        elif &#34;I&#34; in cig:
            start += int(cig[:-1])
        elif any(x in cig for x in [&#34;D&#34;, &#34;N&#34;]):
            sequence_ignored.append(&#34;N&#34; * int(cig[:-1]))
    realignment = deepcopy(alignment)
    realignment[&#34;SEQ&#34;] = &#34;&#34;.join(sequence_ignored)
    return realignment</code></pre>
</details>
</dd>
<dt id="midsv.format.remove_resequence"><code class="name flex">
<span>def <span class="ident">remove_resequence</span></span>(<span>samdict: list[list]) ‑> list[list]</span>
</code></dt>
<dd>
<div class="desc"><p>Remove non-microhomologic overlapped reads within the same QNAME.
The overlapped sequences can be (1) realignments by microhomology or (2) resequence by sequencing error.
The resequenced reads are overlapped but not the same DNA sequence. In contrast, realignments preserve the same sequence.
Since resequence is a Nanopore sequencing error, the reads should be discarded.
In addition, if the start/end of two reads are exactly same, these sequences are resequenced so discarted.
Example reads are in <code>tests/data/overlap/real_overlap.sam</code> and <code>tests/data/overlap/real_overlap2.sam</code></p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>disctionalized SAM</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>list[list]</code></dt>
<dd>disctionalized SAM with removed overlaped reads</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def remove_resequence(samdict: list[list]) -&gt; list[list]:
    &#34;&#34;&#34;Remove non-microhomologic overlapped reads within the same QNAME.
    The overlapped sequences can be (1) realignments by microhomology or (2) resequence by sequencing error.
    The resequenced reads are overlapped but not the same DNA sequence. In contrast, realignments preserve the same sequence.
    Since resequence is a Nanopore sequencing error, the reads should be discarded.
    In addition, if the start/end of two reads are exactly same, these sequences are resequenced so discarted.
    Example reads are in `tests/data/overlap/real_overlap.sam` and `tests/data/overlap/real_overlap2.sam`

    Args:
        sam (list[list]): disctionalized SAM

    Returns:
        list[list]: disctionalized SAM with removed overlaped reads
    &#34;&#34;&#34;
    samdict.sort(key=lambda x: x[&#34;QNAME&#34;])
    sam_groupby = groupby(samdict, lambda x: x[&#34;QNAME&#34;])
    sam_nonoverlapped = []
    for _, alignments in sam_groupby:
        alignments = list(alignments)
        if len(alignments) == 1:
            sam_nonoverlapped += alignments
            continue
        alignments = [realign_sequence(alignment) for alignment in alignments]
        alignments = sorted(alignments, key=lambda x: [x[&#34;POS&#34;], len(x[&#34;SEQ&#34;])]) # real_overlap2 needs comparisons in the following order: (1) shortest read (2) longest read ,(3) middle read
        is_overraped = False
        end_of_previous_read = -1
        previous_read = alignments[0][&#34;SEQ&#34;]
        for alignment in alignments:
            start_of_current_read = alignment[&#34;POS&#34;] - 1
            if end_of_previous_read &gt; start_of_current_read:
                end_of_current_read = return_end_of_current_read(alignment)
                start_overlap = max(start_of_previous_read, start_of_current_read)
                end_overlap = min(end_of_previous_read, end_of_current_read)
                if start_of_previous_read == start_of_current_read:
                    is_overraped = True
                    break
                if end_of_previous_read == end_of_current_read:
                    is_overraped = True
                    break
                for prev, curr in zip(previous_read[start_overlap: end_overlap], alignment[&#34;SEQ&#34;][start_overlap: end_overlap]):
                    if prev == &#34;N&#34; or curr == &#34;N&#34;:
                        continue
                    if prev != curr:
                        is_overraped = True
                        break
            end_of_previous_read = return_end_of_current_read(alignment)
            start_of_previous_read = start_of_current_read
        if not is_overraped:
            sam_nonoverlapped += alignments
    return sam_nonoverlapped</code></pre>
</details>
</dd>
<dt id="midsv.format.remove_softclips"><code class="name flex">
<span>def <span class="ident">remove_softclips</span></span>(<span>sam: list[dict]) ‑> list[dict]</span>
</code></dt>
<dd>
<div class="desc"><p>Remove softclip information from SEQ and QUAL.</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>sam</code></strong> :&ensp;<code>list[list]</code></dt>
<dd>disctionalized SAM</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>list[list]</code></dt>
<dd>disctionalized SAM with trimmed softclips in QUAL</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def remove_softclips(sam: list[dict]) -&gt; list[dict]:
    &#34;&#34;&#34;Remove softclip information from SEQ and QUAL.

    Args:
        sam (list[list]): disctionalized SAM

    Returns:
        list[list]: disctionalized SAM with trimmed softclips in QUAL
    &#34;&#34;&#34;
    sam_list = []
    for alignment in sam:
        cigar = alignment[&#34;CIGAR&#34;]
        if &#34;S&#34; not in cigar:
            sam_list.append(alignment)
            continue
        cigar_split = split_cigar(cigar)
        left, right = cigar_split[0], cigar_split[-1]
        if &#34;S&#34; in left:
            left = int(left[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][left:]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][left:]
        if &#34;S&#34; in right:
            right = int(right[:-1])
            alignment[&#34;SEQ&#34;] = alignment[&#34;SEQ&#34;][:-right]
            alignment[&#34;QUAL&#34;] = alignment[&#34;QUAL&#34;][:-right]
        sam_list.append(alignment)
    return sam_list</code></pre>
</details>
</dd>
<dt id="midsv.format.return_end_of_current_read"><code class="name flex">
<span>def <span class="ident">return_end_of_current_read</span></span>(<span>alignment: dict) ‑> int</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def return_end_of_current_read(alignment:dict) -&gt; int:
    start_of_current_read = alignment[&#34;POS&#34;]
    cigar = alignment[&#34;CIGAR&#34;]
    cigar_split = split_cigar(cigar)
    alignment_length = 0
    for cig in cigar_split:
        if &#34;M&#34; in cig or &#34;D&#34; in cig or &#34;N&#34; in cig:
            alignment_length += int(cig[:-1])
    return start_of_current_read + alignment_length - 1</code></pre>
</details>
</dd>
<dt id="midsv.format.split_cigar"><code class="name flex">
<span>def <span class="ident">split_cigar</span></span>(<span>cigar: str) ‑> list[str]</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def split_cigar(cigar: str) -&gt; list[str]:
    cigar_iter = iter(re.split(r&#34;([MIDNSHPX=])&#34;, cigar))
    cigar_splitted = [i + op for i, op in zip(cigar_iter, cigar_iter)]
    return cigar_splitted</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="midsv" href="index.html">midsv</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="midsv.format.dictionarize_sam" href="#midsv.format.dictionarize_sam">dictionarize_sam</a></code></li>
<li><code><a title="midsv.format.extract_sqheaders" href="#midsv.format.extract_sqheaders">extract_sqheaders</a></code></li>
<li><code><a title="midsv.format.realign_sequence" href="#midsv.format.realign_sequence">realign_sequence</a></code></li>
<li><code><a title="midsv.format.remove_resequence" href="#midsv.format.remove_resequence">remove_resequence</a></code></li>
<li><code><a title="midsv.format.remove_softclips" href="#midsv.format.remove_softclips">remove_softclips</a></code></li>
<li><code><a title="midsv.format.return_end_of_current_read" href="#midsv.format.return_end_of_current_read">return_end_of_current_read</a></code></li>
<li><code><a title="midsv.format.split_cigar" href="#midsv.format.split_cigar">split_cigar</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>